File: 0120 - Triangle/solution.py
class Solution:
    def minimumTotal(self, A: List[List[int]]) -> int:
        @lru_cache(None)
        def dp(i, j):
            if i == 0:
                return A[0][0]
            if j < 0 or j > i:
                return inf
            return A[i][j]+min(dp(i-1, j-1), dp(i-1, j))
        n = len(A)
        return min(dp(n-1, j) for j in range(n))

File: 0120 - Triangle/solution.c
#define INF 0x3f3f3f3f
#define min(a,b) (((a)<(b))?(a):(b))
int minimumTotal(int** A, int sz, int* colSz){
    for(int i = 1; i < sz; i++)
        for(int j = 0; j <= i; j++)
            A[i][j] += min(j!=0?A[i-1][j-1]:INF, j!=i?A[i-1][j]:INF);
    int res = INF;
    for(int i = 0; i < sz; i++) res = min(res, A[sz-1][i]);
    return res;
}